server/auth/types/basic/routes.ts

```diff
@@ -0,0 +1,271 @@
+          let preferredTenants = this.config.multitenancy.tenants.preferred;
+
+          // TODO: figureout selected tenant here and set it in the cookie
+
+          return response.ok({
+            body: {
+              username: user.username,
+              tenants: user.tenants,
+              roles: user.roles,
+              backendroles: user.backendRoles,
+              selectedTenants: '', // TODO: determine selected tenants
+            }
+          })
+        }
+        return response.ok({
+          body: {
+            username: user.username,
+            tenants: user.tenants,
+          }
+        });
+      },
+    );
+
+    // logout
+    this.router.post({
+      path: `${PREFIX}/auth/logout`,
+      validate: false,
+      options: {
+        authRequired: false,
+      }
+    },
+    async (context, request, response) => {
+      this.sessionStorageFactory.asScoped(request).clear();
+      return response.ok(); // TODO: redirect to login?
+    });
+
+    // anonymous auth
+    this.router.get({
+      path: `${PREFIX}/auth/anonymous`,
+      validate: false,
+      options: {
+        authRequired: false,
+      },
+    },
+    async (context, request, response) => {
+      if (this.config.auth.anonymous_auth_enabled) {
+        // TODO: implement anonymous auth for basic authentication
+      } else {
+        return response.redirected({
+          headers: {
+            location: `${PREFIX}/login`,
+          }
+        })
+      }
+    });
+
+    // renders custom error page
+    this.router.get({
+      path: `${PREFIX}/customerror`,
+      validate: false,
+      options: {
+        authRequired: false,
+      }
+    },
+    async (context, request, response) => {
+      return response.ok({
+        body: '',
+      })
+    });
+  }
+
+  // session storage plugin's authenticateWithHeaders() function
+  private async authenticateWithHeaders(request: KibanaRequest, credentials: any = {}, options: any = {}) {
+    try {
+      const additionalAuthHeaders = filterAuthHeaders(request.headers, this.authConfig.allowedAdditionalAuthHeaders);
+      let user = await this.securityClient.authenticateWithHeaders(request, credentials, additionalAuthHeaders);
+
+      let session: SecuritySessionCookie = {
+        username: user.username,
+        credentials: credentials,
+        authType: this.authConfig.authType,
+        assignAuthHeader: false,
+      };
+      let sessionTtl = this.config.session.ttl;
+      if (sessionTtl) {
+        session.expiryTime = Date.now() + sessionTtl;
+      }
+      const authResponse: AuthResponse = {
+        session,
+        user,
+      };
+
+      return this._handleAuthResponse(request, authResponse, additionalAuthHeaders);
+    } catch (error) {
+      this.sessionStorageFactory.asScoped(request).clear();
+      throw error;
+    }
+  }
+
+  private _handleAuthResponse(request: KibanaRequest, authResponse: AuthResponse, additionalAuthHeaders: any = {}) {
```

Comment:
np: can we remove `_`  prefix as the method is private.

Reply:
removed prefix '-'